Make PWM work with Mega boards

photodude opened the pull request:
Timer one is on a different set of pins for Mega boards from Uno/Leonardo/Due Timer one pins
This PR makes PWM on timer one pins (11,12) work for Mega boards, **without breaking backward compatibility** for Uno/Leonardo/Due Timer one pins (9,10)

I have tested this change on a Mega and an Uno board with a modified version of this Library I'm working on for the Single G2 breakout boards (needed a library for the single boards for a project I'm working on) 

https://github.com/photodude/G2MotorDriver
If you're interested, I would be willing to transfer this modified library to pololu in a month or so when I'm no longer modifying the library with changes related to testing from my project.

photodude committed:
Make PWM work with Mega boards

Timer one is on a different set of pins for Mega boards. 
This makes PWM on timer one pins (11,12) work for Mega boards, without breaking backward compatibility for Uno/Leonardo/Due Timer one pins (9,10)

photodude commented:
@DavidEGrayson @ryantm @kevin-pololu bump for review and merge consideration

photodude commented:
@DavidEGrayson @ryantm @kevin-pololu bump for review and merge consideration, Please let me know if you would be interested in adopting this change.

ryantm commented:
Hi, @photodude.

Sorry for the extremely late reply. I'm wary about this change, especially without documentation, because it wouldn't work if you plugged the board into an Arduino Mega. Also, it uses pin 12, which is the default pin for M2EN/DIAG without changing that pin's assignment.

It might make more sense to rename `DUALG2HIGHPOWERMOTORSHIELD_TIMER1_AVAILABLE` to `DUALG2HIGHPOWERMOTORSHIELD_TIMER1_AVAILABLE_ON_9_AND_10`, then introduce another definition called `DUALG2HIGHPOWERMOTORSHIELD_TIMER2_AVAILABLE_ON_9_AND_10`, which the Mega could use by default.

Sincerely,
Ryan Mulligan.

photodude committed:
Revised Mega PWM approch

Revised Mega PWM approach based on suggestion from @ryantm

ICR2 should be verified before uncommenting.

photodude committed:
Revised Mega timer adjustments

Remember to set the OCR1A/OCR1B vs OCR2A/OCR2B depending on which timer the board uses.

photodude commented:
@ryantm Based on your suggestion I have followed your approach to use the 8 bit timer on pin 9 and 10 for the Mega PWM change. This should now maintain the plug-n-play setting for the hardware. Although the PWM signal on the mega is lower frequency due to the 8 bit timer2 vs 16 bit timer1; it's superior to the noisy `analogWrite()` pseudo PWM

I wasn't 100% confident in the `ICR2` value in the PWM calculation for Timer 2 on the mega so I left that commented out.

Let me know if any additional documentation or commenting is needed.

photodude commented:
@ryantm What Documentation or additional commenting should be provided with this change?

photodude commented:
@ryantm What Documentation or additional commenting should be provided with this change?
Can we work towards Merge or shall this be closed?

DavidEGrayson commented:
I think you got the pins backwards.  According to an official [pin mapping document](https://docs.arduino.cc/hacking/hardware/PinMapping2560):

- Digital pin 9 = PH6 = OC2B
- Digital pin 10 = PB4 = OC2A

The mapping in your code is different:

- Digital pin 9 = _M1PWM_TIMER_PIN, and setM1Speed uses OCR2A
- Digital pin 10 = _M2PWM_TIMER_PIN, and setM2Speed uses OCR2B

Also, defining macros in DualG2HighPowerMotorShield.h that do not have an appropriate prefix is dangerous because they might collide with constants from similar libraries.  I think those definitions should be moved to the CPP file.

I have not carefully checked the register settings you are using.

In terms of documentation, we mention Timer1 in the reference section of the README.md but accepting your pull request as is would make those comments outdated.  I think the documentation of the constructor should probably say under what conditions a timer will be used.  For example:  "On boards based on the ATmega1280 or ATmega2560, if M1PWM is ___ and M2PWM is ___, the this library will generate PWM using Timer 2."  There is also some discussion about PWM frequency and analogWrite in the "Current readings" section that should be updated, but we should avoid duplicating a lot of information in multiple places in the documentation.

photodude commented:
@DavidEGrayson I'll have to check my notes but the call here is not to Pin 9 OC2B and is not using the "PWM on a subset of the Mega output pins" but to the Registry timer OC**R**2A
Off the top of my head I think that my reference for writing and testing this was [arduino documentation Secrets to PWM](https://docs.arduino.cc/tutorials/generic/secrets-of-arduino-pwm) and the "fast PWM" or Phase-Correct PWM implementation. Which also follows how the library is currently written.

I believe it's "OC**R**2A as the top value for the timer."... "and the OC**R**2B compare register" or maybe the library is using both as "compare register"
and there's some other fancy stuff with TCCR2A, TCCR2B for frequency calculation

DavidEGrayson commented:
Your code writes a motor speed to OCR2A in setM1Speed so that register is not being used as the TOP value.

photodude commented:
@DavidEGrayson I may have the specifics off, but that's also what the existing code does. https://github.com/pololu/dual-g2-high-power-motor-shield/blob/master/DualG2HighPowerMotorShield.cpp#L94
